upup/pkg/fi/cloudup/hetzner/cloud.go

```diff
@@ -0,0 +1,523 @@
+		if cluster.Spec.MasterPublicName != "" {
+			// Note that this must match Digital Ocean's lb name
+			klog.V(2).Infof("Querying Hetzner to find Loadbalancers for API (%q)", cluster.Name)
+
+			loadBalancers, err := c.GetAllLoadBalancers()
+			if err != nil {
+				return false, fmt.Errorf("LoadBalancers.List returned error: %v", err)
+			}
+
+			lbName := "api." + cluster.Name
+
+			for _, lb := range loadBalancers {
+				if lb.Name == lbName {
+					klog.V(2).Infof("Matching LB name found for API (%q)", cluster.Name)
+
+					if !lb.PublicNet.Enabled {
+						return false, fmt.Errorf("load-balancer is not yet active")
+					}
+
+					address := lb.PublicNet.IPv4.IP.String()
+					ingresses = append(ingresses, fi.ApiIngressStatus{IP: address})
+				}
+			}
+		}
+		return true, nil
+	})
+	if done {
+		return ingresses, nil
+	} else {
+		if err == nil {
+			err = wait.ErrWaitTimeout
+		}
+		return ingresses, err
+	}
+}
+
+// FindClusterStatus discovers the status of the cluster, by looking for the tagged etcd volumes
+func (c *hetznerCloudImplementation) FindClusterStatus(cluster *kops.Cluster) (*kops.ClusterStatus, error) {
+	////etcdStatus, err := findEtcdStatus(c, cluster)
+	////if err != nil {
+	////	return nil, err
+	////}
+	//status := &kops.ClusterStatus{
+	//	EtcdClusters: etcdStatus,
+	//}
+	//klog.V(2).Infof("Cluster status (from cloud): %v", fi.DebugAsJsonString(status))
+	status := &kops.ClusterStatus{}
+	return status, nil
+}
+
+func getCloudGroups(c *hetznerCloudImplementation, cluster *kops.Cluster, instancegroups []*kops.InstanceGroup, warnUnmatched bool, nodes []v1.Node) (map[string]*cloudinstances.CloudInstanceGroup, error) {
+	nodeMap := cloudinstances.GetNodeMap(nodes, cluster)
+
+	groups := make(map[string]*cloudinstances.CloudInstanceGroup)
+	instanceGroups, err := findInstanceGroups(c, cluster.ObjectMeta.Name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find autoscale groups: %v", err)
+	}
+
+	for _, doGroup := range instanceGroups {
+		name := doGroup.InstanceGroupName
+
+		instancegroup, err := matchInstanceGroup(name, cluster.ObjectMeta.Name, instancegroups)
+		if err != nil {
+			return nil, fmt.Errorf("error getting instance group for doGroup %q", name)
+		}
+		if instancegroup == nil {
+			if warnUnmatched {
+				klog.Warningf("Found doGroup with no corresponding instance group %q", name)
+			}
+			continue
+		}
+
+		groups[instancegroup.ObjectMeta.Name], err = buildCloudInstanceGroup(c, instancegroup, doGroup, nodeMap)
+		if err != nil {
+			return nil, fmt.Errorf("error getting cloud instance group %q: %v", instancegroup.ObjectMeta.Name, err)
+		}
+	}
+
+	klog.V(8).Infof("Cloud Instance Group Info = %v", groups)
+	return groups, nil
+}
+
+// findInstanceGroups finds instance groups matching the specified tags
+func findInstanceGroups(c *hetznerCloudImplementation, clusterName string) ([]HetznerInstanceGroup, error) {
+	var result []HetznerInstanceGroup
```

Comment:
Nit: Add a TODO.  I think we can implement something like this by grouping instances by their instance-group label.  Maybe?